presto-main/src/main/java/com/facebook/presto/sql/planner/NodePartitioningManager.java

```diff
@@ -258,9 +258,25 @@ private ConnectorBucketNodeMap getConnectorBucketNodeMap(Session session, Partit
         return distribution.build();
     }
 
-    public List<Node> sortedNodes(Session session, ConnectorId connectorId)
+    public List<Node> getNodes(Session session, ConnectorId connectorId)
     {
-        List<InternalNode> nodes = nodeScheduler.createNodeSelector(session, connectorId).allNodes();
-        return nodes.stream().sorted(comparing(InternalNode::getNodeIdentifier)).collect(toImmutableList());
+        // Nodes returned by the node selector are already sorted based on nodeIdentifier. No need to sort again
+        List<InternalNode> allNodes = nodeScheduler.createNodeSelector(session, connectorId).getAllNodes();
+
+        ImmutableList.Builder<Node> nodeBuilder = ImmutableList.builder();
+        int nodeCount = allNodes.size();
+        for (int i = 0; i < nodeCount; i++) {
+            InternalNode node = allNodes.get(i);
+            if (node.getNodeStatus() == DEAD) {
+                // Replace dead nodes with the first alive node to the right of it in the sorted node list
```

Comment:
I am not sure if this is the correct logic. With last node being dead, this will resume the loop again. In this case it would add node 0 two times. Why dont we just filter out dead nodes? Why do we need to maintain index?